lib/vnet/daemon/service_darwin.m

```diff
@@ -0,0 +1,147 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "common_darwin.h"
+#include "protocol_darwin.h"
+#include "service_darwin.h"
+
+#import <Foundation/Foundation.h>
+
+#include <string.h>
+
+@interface VNEDaemonService () <NSXPCListenerDelegate, VNEDaemonProtocol>
+
+@property(nonatomic, strong, readwrite) NSXPCListener *listener;
+@property(nonatomic, readwrite) BOOL started;
+
+@property(nonatomic, readwrite) NSString *socketPath;
+@property(nonatomic, readwrite) NSString *ipv6Prefix;
+@property(nonatomic, readwrite) NSString *dnsAddr;
+@property(nonatomic, readwrite) NSString *homePath;
+@property(nonatomic, readwrite) dispatch_semaphore_t gotVnetConfigSema;
+
+@end
+
+@implementation VNEDaemonService
+
+- (id)initWithBundlePath:(NSString *)bundlePath {
+  self = [super init];
+  if (self) {
+    // Launch daemons must configure their listener with the machServiceName
+    // initializer.
+    _listener = [[NSXPCListener alloc] initWithMachServiceName:DaemonLabel(bundlePath)];
+    _listener.delegate = self;
+
+    _started = NO;
+    _gotVnetConfigSema = dispatch_semaphore_create(0);
+  }
+  return self;
+}
+
+- (void)start {
+  assert(_started == NO);
+
+  // Begin listening for incoming XPC connections.
+  [_listener resume];
+
+  _started = YES;
+}
+
+- (void)stop {
+  assert(_started == YES);
+
+  // Stop listening for incoming XPC connections.
+  [_listener suspend];
+
+  _started = NO;
+  dispatch_semaphore_signal(_gotVnetConfigSema);
+}
+
+- (void)waitForVnetConfig {
+  assert(_started == YES);
+  dispatch_semaphore_wait(_gotVnetConfigSema, DISPATCH_TIME_FOREVER);
+}
+
+#pragma mark - VNEDaemonProtocol
+
+- (void)startVnet:(VnetConfig *)vnetConfig completion:(void (^)(void))completion {
+  @synchronized(self) {
+    _socketPath = @(vnetConfig->socket_path);
+    _ipv6Prefix = @(vnetConfig->ipv6_prefix);
+    _dnsAddr = @(vnetConfig->dns_addr);
+    _homePath = @(vnetConfig->home_path);
+    dispatch_semaphore_signal(_gotVnetConfigSema);
+    completion();
+  }
+}
+
+#pragma mark - NSXPCListenerDelegate
+
+- (BOOL)listener:(NSXPCListener *)listener
+    shouldAcceptNewConnection:(NSXPCConnection *)newConnection {
+  assert(listener == _listener);
+  assert(newConnection != nil);
+
+  // Configure the incoming connection.
+  newConnection.exportedInterface =
+      [NSXPCInterface interfaceWithProtocol:@protocol(VNEDaemonProtocol)];
+  newConnection.exportedObject = self;
+
+  // New connections always start in a suspended state.
+  [newConnection resume];
+
+  return YES;
+}
+
+@end
+
+static VNEDaemonService *daemonService = NULL;
+
+void DaemonStart(const char *bundle_path) {
+  daemonService = [[VNEDaemonService alloc] initWithBundlePath:@(bundle_path)];
```

Comment:
Are we OK with an eventual double-initialization here? Should we check for that, error, etc?

Reply:
I'll add a noop if the daemon was already initialized or started. In this version, I added the checks to `DaemonStop` only since it's called sort of as a "clean up".